### PR TITLE
Consensus fix, don't consider threshold check if..

### LINKED ISSUE #7904: SLP 1.1 has stalled
Metrics:
* When the tower flatline happened: https://metrics.solana.com:3000/d/testnet/testnet-monitor?orgId=2&from=1579639842339&to=1579639922310&var-testnet=cluster&var-hostid=All
* ~20 minutes later when InvalidEntryHash blocktree errors started to occur:  https://metrics.solana.com:3000/d/testnet/testnet-monitor?orgId=2&from=1579641230438&to=1579641530000&var-testnet=cluster&var-hostid=All&panelId=34

Looks like we had a bad hash (InvalidEntryHash) at or around slot 1033138, which caused a loss of consensus

### PR DESCRIPTION
**Pull request by sakridge**
### Problem

Nodes stop voting because of threshold when they are trying to switch back to the main heavier fork after voting on a minor fork. This stalls the network since all nodes do not reach consensus on the heavier fork.

#### Summary of Changes

Do not consider threshold if the vote does not affect lockouts in the new tower state.

Fixes #7904

### CODE REVIEW

File: core/src/consensus.rs [hunk 1]
```diff
@@ -321,13 +321,24 @@ impl Tower {
...
-                    "fork_stake {} {} {} {}",
+                    "fork_stake slot: {} lockout: {} fork_stake: {} total_stake: {}",
                     slot,
                     lockout,
                     fork_stake.stake,
                     total_staked
                 );
-                lockout > self.threshold_size
+                for (new_lockout, original_lockout) in
+                    lockouts.votes.iter().zip(self.lockouts.votes.iter())
```
Comment (carllin) [inline core/src/consensus.rs]: 1) I think we should only care about lockouts doubling if they are beyond the nth vote
2) The `process_slot_vote_unchecked` could have popped off some votes. I think if it pops off any votes then there's no need to check threshold either, as I don't think you can pop off votes and also double lockout

Comment (sakridge) [reply core/src/consensus.rs]: For 1. if popping off results in not doubling lockout, then why not just test for not doubling lockout which we should do anyway even if we don't pop.

For 2, yea I will bail early if we make it to a threshold slot.

File: core/src/partition_cfg.rs [hunk 2]
```diff
@@ -12,33 +12,49 @@ use std::sync::RwLock;
...
 
 impl PartitionCfg {
     pub fn new(partitions: Vec<Partition>) -> Self {
-        Self { partitions }
+        Self {
+            partitions,
+            printed: HashSet::new(),
+        }
+    }
+    pub fn all_partitions_done(&self) -> bool {
```
Comment (carllin) [inline core/src/partition_cfg.rs]: looks like this can be removed

Comment (sakridge) [reply core/src/partition_cfg.rs]: this whole file could be removed I think.

**Pull request merged**